File: Software/Chapter 14/vb/Listing14-2/Listing14-2/Listing14-2.vb
Imports System
Imports System.Data
Imports System.Data.SqlClient
Imports System.Data.SqlTypes
Imports Microsoft.SqlServer.Server
Imports System.Xml
Imports System.Runtime.InteropServices

Namespace APress.Samples
    Partial Public Class Sql
        <SqlFunction(IsDeterministic:=False, _
            DataAccess:=DataAccessKind.None, _
            TableDefinition:="title nvarchar(256), link nvarchar(256), " & _
                "pubdate datetime, description nvarchar(max)", _
            FillRowMethodName:="GetRow")> _
        Public Shared Function GetYahooNews() As IEnumerable
            Dim xmlsource As New _
                XmlTextReader("http://rss.news.yahoo.com/rss/topstories")
            Dim newsxml As New XmlDocument
            newsxml.Load(xmlsource)
            xmlsource.Close()
            Return newsxml.SelectNodes("//rss/channel/item")
        End Function

        Private Shared Sub GetRow(ByVal o As Object, _
            <Out()> ByRef title As SqlString, _
            <Out()> ByRef link As SqlString, _
            <Out()> ByRef pubdate As SqlDateTime, _
            <Out()> ByRef description As SqlString)
            Dim element As XmlElement
            element = DirectCast(o, XmlElement)
            title = element.SelectSingleNode("./title").InnerText
            link = element.SelectSingleNode("./link").InnerText
            pubdate = CType(element.SelectSingleNode("./pubDate").InnerText, _
                DateTime)
            description = element.SelectSingleNode("./description").InnerText
        End Sub
    End Class
End Namespace

File: Software/Chapter 14/cs/Listing14-2/Listing14-2/Listing14-2.cs
using System;
using System.Data;
using System.Data.SqlClient;
using System.Data.SqlTypes;
using Microsoft.SqlServer.Server;
using System.Xml;
using System.Runtime.InteropServices;
using System.Collections;

namespace APress.Samples
{

    public partial class Sql
    {
        [SqlFunction(IsDeterministic = false,
            DataAccess = DataAccessKind.None,
            TableDefinition = "title nvarchar(256), link nvarchar(256), " +
                "pubdate datetime, description nvarchar(max)",
            FillRowMethodName = "GetRow")]
        public static IEnumerable GetYahooNews()
        {
            XmlTextReader xmlsource = new XmlTextReader("http://rss.news.yahoo.com/rss/topstories");
            XmlDocument newsxml = new XmlDocument();
            newsxml.Load(xmlsource);
            xmlsource.Close();
            return newsxml.SelectNodes("//rss/channel/item");
        }

        private static void GetRow(Object o, out SqlString title, out SqlString link,
            out SqlDateTime pubdate, out SqlString description)
        {
            XmlElement element;
            element = (XmlElement)o;
            title = element.SelectSingleNode("./title").InnerText;
            link = element.SelectSingleNode("./link").InnerText;
            pubdate = DateTime.Parse(element.SelectSingleNode("./pubDate").InnerText);
            description = element.SelectSingleNode("./description").InnerText;
        }
    };
}